module/lifecycle/lifecycle_test.go

```diff
@@ -2,15 +2,119 @@ package lifecycle_test
 
 import (
 	"fmt"
+	"sync/atomic"
 	"testing"
 	"time"
 
+	"github.com/stretchr/testify/suite"
+
 	realmodule "github.com/onflow/flow-go/module"
 	"github.com/onflow/flow-go/module/lifecycle"
 	module "github.com/onflow/flow-go/module/mock"
 	"github.com/onflow/flow-go/utils/unittest"
 )
 
+type LifecycleManagerSuite struct {
+	suite.Suite
+	lm *lifecycle.LifecycleManager
+}
+
+func (suite *LifecycleManagerSuite) SetupTest() {
+	suite.lm = lifecycle.NewLifecycleManager()
+}
+
+// TestConsecutiveStart tests that calling OnStart multiple times concurrently only
+// results in startup being performed once
+func (suite *LifecycleManagerSuite) TestConsecutiveStart() {
+	var numStarts uint32
+
+	for i := 0; i < 10; i++ {
+		go func() {
+			suite.lm.OnStart(func() {
+				atomic.AddUint32(&numStarts, 1)
+			})
+		}()
+	}
+
+	unittest.RequireCloseBefore(suite.T(), suite.lm.Started(), time.Second, "timed out waiting for startup")
+	time.Sleep(100 * time.Millisecond) // wait for potential race conditions to occur
+
+	suite.Assert().EqualValues(1, numStarts)
+}
+
+// TestConsecutiveStop tests that calling OnStop multiple times concurrently only
+// results in shutdown being performed once
+func (suite *LifecycleManagerSuite) TestConsecutiveStop() {
+	suite.lm.OnStart()
+	unittest.RequireCloseBefore(suite.T(), suite.lm.Started(), time.Second, "timed out waiting for startup")
+
+	var numStops uint32
+
+	for i := 0; i < 10; i++ {
+		go func() {
+			suite.lm.OnStop(func() {
+				atomic.AddUint32(&numStops, 1)
+			})
+		}()
+	}
+
+	unittest.RequireCloseBefore(suite.T(), suite.lm.Stopped(), time.Second, "timed out waiting for shutdown")
+	time.Sleep(100 * time.Millisecond) // wait for potential race conditions to occur
+
+	suite.Assert().EqualValues(numStops, 1)
+}
+
+// TestStopBeforeStart tests that calling OnStop before OnStart results in startup never
+// being performed, and the returned channel never closing.
+func (suite *LifecycleManagerSuite) TestStopBeforeStart() {
+	suite.lm.OnStop(func() {
+		suite.FailNow("shutdown should not occur")
+	})
+
+	suite.lm.OnStart(func() {
+		suite.FailNow("startup should not occur")
+	})
+
+	unittest.RequireCloseBefore(suite.T(), suite.lm.Stopped(), time.Second, "timed out waiting for shutdown")
+	unittest.RequireNotClosed(suite.T(), suite.lm.Started(), "Started channel should never close")
+}
+
+// TestStopAfterStart tests that if OnStop is called right after OnStart, shutdown will
+// only be performed after startup has finished.
+func (suite *LifecycleManagerSuite) TestStopAfterStart() {
+	var started uint32 = 0
+
+	suite.lm.OnStart(func() {
+		// simulate startup processing
+		time.Sleep(3 * time.Second)
```

Comment:
**General comment: ** sleep durations in this PR are in order of seconds. This slows down building over CI. Would be great if you could please reduce them to milliseconds if possible. For most of the cases, `100` milliseconds is a preferable sleep duration

Reply:
or you could use a channel here as a gate. After the call to `suite.lm.OnStop()`, you can close the channel (a.k.a. open the gate). Then we won't have to rely on a `sleep`